### PR TITLE
Ensure all legacy sharding commands are clearly deprecated

### PR DESCRIPTION
**Pull request by mattlord**
## Description

The legacy sharding code — e.g. `SplitClone` — has long since been replaced by VReplication based workflows like `MoveTables`.

This PR ensures that all related user commands are clearly marked as deprecated so that all related code can be removed in v15.

## Related Issue(s)

- https://github.com/vitessio/vitess/issues/9219
- https://github.com/vitessio/vitess/pull/10278

## Checklist

-   [x] "Backport me!" label has been added if this change should be backported
-   [x] Tests are not required
-   [x] Documentation: https://github.com/vitessio/website/pull/1027

### CODE REVIEW

File: go/vt/binlog/keyspace_id_resolver.go [hunk 1]
```diff
@@ -34,7 +34,7 @@ import (
 	topodatapb "vitess.io/vitess/go/vt/proto/topodata"
 )
 
-var useV3ReshardingMode = flag.Bool("binlog_use_v3_resharding_mode", true, "True iff the binlog streamer should use V3-style sharding, which doesn't require a preset sharding key column.")
+var useV3ReshardingMode = flag.Bool("binlog_use_v3_resharding_mode", true, "(DEPRECATED) True if the binlog streamer should use V3-style sharding, which doesn't require a preset sharding key column.")
```
Comment (deepthi) [inline go/vt/binlog/keyspace_id_resolver.go:37]: I think the `iff` was intentional - meaning `if-and-only-if`. But I'm fine with removing the whole description and just keeping `DEPRECATED`. Whatever change you make here needs to go into flags_test.go as well.

File: go/vt/vtctl/vtctl.go [hunk 2]
```diff
@@ -332,16 +332,16 @@ var commands = []commandGroup{
 				name:   "SetShardTabletControl",
 				method: commandSetShardTabletControl,
 				params: "[--cells=c1,c2,...] [--denied_tables=t1,t2,...] [--remove] [--disable_query_service] <keyspace/shard> <tablet type>",
-				help: "Sets the TabletControl record for a shard and type. Only use this for an emergency fix or after a finished vertical split. The *MigrateServedFrom* and *MigrateServedType* commands set this field appropriately already. Always specify the denied_tables flag for vertical splits, but never for horizontal splits.\n" +
-					"To set the DisableQueryServiceFlag, keep 'denied_tables' empty, and set 'disable_query_service' to true or false. Useful to fix horizontal splits gone wrong.\n" +
-					"To change the list of denied tables, specify the 'denied_tables' parameter with the new list. Useful to fix tables that are being blocked after a vertical split.\n" +
-					"To just remove the ShardTabletControl entirely, use the 'remove' flag, useful after a vertical split is finished to remove serving restrictions.",
+				help: "Sets the TabletControl record for a shard and type. Only use this for an emergency fix.\n" +
```
Comment (deepthi) [inline go/vt/vtctl/vtctl.go:335]: type -> tablet type

File: go/vt/vtctl/vtctl.go [hunk 3]
```diff
@@ -2820,6 +2831,7 @@ func commandMigrateServedTypes(ctx context.Context, wr *wrangler.Wrangler, subFl
 }
 
 func commandMigrateServedFrom(ctx context.Context, wr *wrangler.Wrangler, subFlags *flag.FlagSet, args []string) error {
+	wr.Logger().Printf("*** This is a legacy sharding command that will soon be removed, please use VReplication instead https://vitess.io/docs/reference/vreplication/ ***\n")
```
Comment (deepthi) [inline go/vt/vtctl/vtctl.go:2834]: This is the one place where you didn't use an `!`. Do you wanna make it consistent with the other warnings?

File: go/vt/worker/grpcvtworkerclient/client.go [hunk 4]
```diff
@@ -36,11 +36,11 @@ import (
 )
 
 var (
-	cert = flag.String("vtworker_client_grpc_cert", "", "the cert to use to connect")
-	key  = flag.String("vtworker_client_grpc_key", "", "the key to use to connect")
-	ca   = flag.String("vtworker_client_grpc_ca", "", "the server ca to use to validate servers when connecting")
-	crl  = flag.String("vtworker_client_grpc_crl", "", "the server crl to use to validate server certificates when connecting")
-	name = flag.String("vtworker_client_grpc_server_name", "", "the server name to use to validate server certificate")
+	cert = flag.String("vtworker_client_grpc_cert", "", "(DEPRECATED) the cert to use to connect")
```
Comment (deepthi) [inline go/vt/worker/grpcvtworkerclient/client.go:39]: Can we add a deprecation warning when the vtworker binary starts up?

File: go/cmd/vtctldclient/command/keyspaces.go [hunk 5]
```diff
@@ -91,7 +91,7 @@ Otherwise, the keyspace must be empty (have no shards), or returns an error.`,
 	// SetKeyspaceServedFrom makes a SetKeyspaceServedFrom gRPC call to a vtcltd.
 	SetKeyspaceServedFrom = &cobra.Command{
 		Use:                   "SetKeyspaceServedFrom [--source <keyspace>] [--remove] [--cells=<cells>] <keyspace> <tablet_type>",
-		Short:                 "Updates the ServedFromMap for a keyspace manually. This command is intended for emergency fixes; the map is automatically set by MigrateServedTypes. This command does not rebuild the serving graph.",
+		Short:                 "Updates the ServedFromMap for a keyspace manually. This command is intended for emergency fixes. This command does not rebuild the serving graph.",
```
Comment (deepthi) [inline go/cmd/vtctldclient/command/keyspaces.go:94]: We should add `Deprecated: ....` to this one.

**Comment by deepthi** [comment]
@ajm188 can you review this too?

File: go/vt/worker/grpcvtworkerclient/client.go (hunk 4 above)
Comment (mattlord) [reply go/vt/worker/grpcvtworkerclient/client.go:39]: Sure. I'm assuming you mean print it to STDERR/console. I'll add something.

Comment (mattlord) [reply go/vt/worker/grpcvtworkerclient/client.go:39]:
```
$ vtworker
*** This is a legacy sharding tool that will soon be removed! Please use VReplication instead https://vitess.io/docs/reference/vreplication/ ***
F0513 00:23:34.072379   42823 server.go:224] topo_global_server_address must be configured
```

File: go/vt/vtctl/vtctl.go [hunk 6]
```diff
@@ -510,31 +512,33 @@ var commands = []commandGroup{
 			{
 				name:   "VDiff",
 				method: commandVDiff,
-				params: "[-source_cell=<cell>] [-target_cell=<cell>] [-tablet_types=PRIMARY,REPLICA,RDONLY] [-filtered_replication_wait_time=30s] [-max_extra_rows_to_compare=1000] <keyspace.workflow>",
+				params: "[--source_cell=<cell>] [--target_cell=<cell>] [--tablet_types=in_order:RDONLY,REPLICA,PRIMARY] [--filtered_replication_wait_time=30s] [--max_extra_rows_to_compare=1000] <keyspace.workflow>",
```
Comment (mattlord) [inline go/vt/vtctl/vtctl.go:515]: This was changed in https://github.com/vitessio/vitess/pull/10040 but I missed this in the docs PR: https://github.com/vitessio/website/pull/1003

File: examples/compose/vttablet-up.sh [hunk 7]
```diff
@@ -108,58 +108,57 @@ sleep $sleeptime
 
 # Create the cell
 # https://vitess.io/blog/2020-04-27-life-of-a-cluster/
-$VTROOT/bin/vtctlclient -server vtctld:$GRPC_PORT AddCellInfo -root vitess/$CELL -server_address consul1:8500 $CELL || true
+$VTROOT/bin/vtctlclient --server vtctld:$GRPC_PORT AddCellInfo --root vitess/$CELL --server_address consul1:8500 $CELL || true
```
Comment (ajm188) [inline examples/compose/vttablet-up.sh:111]: I _believe_ you want an extra double dash before `AddCellInfo` (I think this prints out [this warning](https://github.com/vitessio/vitess/blob/main/go/internal/flag/flag.go#L92-L94))

File: go/cmd/vtctldclient/command/keyspaces.go [hunk 8]
```diff
@@ -91,15 +91,15 @@ Otherwise, the keyspace must be empty (have no shards), or returns an error.`,
 	// SetKeyspaceServedFrom makes a SetKeyspaceServedFrom gRPC call to a vtcltd.
 	SetKeyspaceServedFrom = &cobra.Command{
 		Use:                   "SetKeyspaceServedFrom [--source <keyspace>] [--remove] [--cells=<cells>] <keyspace> <tablet_type>",
-		Short:                 "Updates the ServedFromMap for a keyspace manually. This command is intended for emergency fixes; the map is automatically set by MigrateServedTypes. This command does not rebuild the serving graph.",
+		Short:                 "(DEPRECATED) Updates the ServedFromMap for a keyspace manually. This command is intended for emergency fixes. This command does not rebuild the serving graph.",
```
Comment (ajm188) [inline go/cmd/vtctldclient/command/keyspaces.go:94]: You should also set the [`Deprecated`](https://pkg.go.dev/github.com/spf13/cobra#Command) field on these commands, which "defines, if this command is deprecated and should print this string when used."

File: examples/compose/vttablet-up.sh (hunk 7 above)
Comment (mattlord) [reply examples/compose/vttablet-up.sh:111]: Yeah:
```
$ vtctlclient --server localhost:15999 AddCellInfo --root zone2 --server_address foo:1
W0513 15:48:13.814964   23983 flag.go:123] Detected a dashed argument after a positional argument. Currently these are treated as posargs that may be parsed by a subcommand, but in the next version of Vitess they will be parsed as top-level flags, which may not be defined, causing errors. To preserve existing behavior, please update your invocation to include a "--" after all top-level flags to continue treating --root as a positional argument.
...
```

File: go/cmd/vtctldclient/command/keyspaces.go (hunk 8 above)
Comment (mattlord) [reply go/cmd/vtctldclient/command/keyspaces.go:94]: Thanks! I didn't notice that it also existed here. Will change now.

File: go/cmd/vtctldclient/command/keyspaces.go [hunk 9]
```diff
@@ -91,10 +91,11 @@ Otherwise, the keyspace must be empty (have no shards), or returns an error.`,
 	// SetKeyspaceServedFrom makes a SetKeyspaceServedFrom gRPC call to a vtcltd.
 	SetKeyspaceServedFrom = &cobra.Command{
 		Use:                   "SetKeyspaceServedFrom [--source <keyspace>] [--remove] [--cells=<cells>] <keyspace> <tablet_type>",
-		Short:                 "Updates the ServedFromMap for a keyspace manually. This command is intended for emergency fixes; the map is automatically set by MigrateServedTypes. This command does not rebuild the serving graph.",
+		Short:                 "Updates the ServedFromMap for a keyspace manually. This command is intended for emergency fixes. This command does not rebuild the serving graph.",
 		DisableFlagsInUseLine: true,
 		Args:                  cobra.ExactArgs(2),
 		RunE:                  commandSetKeyspaceServedFrom,
+		Deprecated:            "and will soon be removed! Please use VReplication instead: https://vitess.io/docs/reference/vreplication",
```
Comment (mattlord) [inline go/cmd/vtctldclient/command/keyspaces.go:98]: You can see what will get printed here: https://github.com/spf13/cobra/blob/9e88759b19cd358e9af6f0d778b9a23990cf7f51/command.go#L785-L787

**Review by ajm188 — APPROVED** [review]

File: go/vt/worker/grpcvtworkerclient/client.go (hunk 4 above)
Comment (deepthi) [reply go/vt/worker/grpcvtworkerclient/client.go:39]: `Please use VReplication instead` -> `Please use MoveTables or Reshard commands instead`
WDYT?

**Review by deepthi — APPROVED** [review]
I left a suggestion for the deprecation warning.
The rest LGTM

File: go/vt/worker/grpcvtworkerclient/client.go (hunk 4 above)
Comment (mattlord) [reply go/vt/worker/grpcvtworkerclient/client.go:39]: Yeah, if that makes more sense to you — being somebody who is familiar with the legacy sharding stuff  — then I'm all for it. I'll make this change shortly. Thanks!

**Pull request merged**